### ISSUE TITLE
Sparsity with SparseDiffTools

### ISSUE THREAD

**Issue by matthieugomez** [JuliaNLSolvers/NLsolve.jl#219]
It would be great to use SparseDifftools in NLSolve, and make it easier for users to work with sparse Jacobians @ChrisRackauckas

**Comment by ChrisRackauckas** [comment]
The place to start is probably by allowing users to pass a color vector and making use of that.

**Comment by matthieugomez** [comment]
It could also use automatic sparsity detection right?

**Comment by ChrisRackauckas** [comment]
It could, be I think we should be a bit careful still before defaulting to it. Cassette is a pretty big dependency, and we don't know everything about the average compile times for using it yet. Standard runtime is around 2 seconds I think, so that might be more of an opt-in thing.

**Comment by pkofod** [comment]
I'll keep it in mind. A usage example (outside of nlsolve, just a simple application of the tools to a function) would be great!

**Comment by matthieugomez** [comment]
One UI could be something like
```julia
function OnceDifferentiable(f!, x0, F0, J0; color = color(J0), autodiff = true)
   if autodiff == :forward
      jac_cache = ForwardColorJacCache(f!,  x0; color = color)
      j!(J, y) = forwarddiff_color_jacobian!(J, f!, y, jac_cache)
   else
      j!(J, y) = finite_difference_jacobian!(J0, f!, y, color=colors)
   end
   OnceDifferentiable(f!, j!, x0, F0, J0)
end
```

**Comment by ChrisRackauckas** [comment]
That matches our plan with DiffEq

**Comment by pkofod** [comment]
What does J0 need to be here?

**Comment by matthieugomez** [comment]
It allows the user to indicate the sparsity structure of the Jacobian. It is not really needed since `SparseDiffTools` allows to get it automatically (but it's not always super fast). Maybe a better syntax could be:
```julia
 OnceDifferentiable(f!, x0, F0; J0 = Float64.(sparse(sparsity!(f!,F0, x0))) color = matrix_colors(J0), autodiff = true)

**Comment by pkofod** [comment]
>  To make `J0` optional:

I'll figure something out, I was just looking for the basic interface for SparseDiffTools. Thanks!

**Comment by ChrisRackauckas** [comment]
Here's a blog post that explains it a bit:

http://www.stochasticlifestyle.com/a-collection-of-jacobian-sparsity-acceleration-tools-for-julia/

**Comment by pkofod** [comment]
Thanks

**Comment by ChrisRackauckas** [comment]
I think @matthieugomez 's suggestion is all that's needed, and then the differentiation calls in the OnceDifferentiable just need to add the colorvec and sparsity arguments using the colorvec and J0 (it separates `sparsity` and `J` because you may want to decompress into a dense matrix, though I assume for NLsolve's purposes you can abstract that away and just make it the same for both).

**Comment by pkofod** [comment]
> I think @matthieugomez 's suggestion is all that's needed, and then the differentiation calls in the OnceDifferentiable just need to add the colorvec and sparsity arguments using the colorvec and J0 (it separates `sparsity` and `J` because you may want to decompress into a dense matrix, though I assume for NLsolve's purposes you can abstract that away and just make it the same for both).

Agreed, though Matthieu's code doesn't actually make it optional, it forces it to be the default, but I will make a version where you can pass the required info.

**Comment by ChrisRackauckas** [comment]
Now sparsity with FiniteDiff.jl

**Comment by pkofod** [comment]
I will switch over in NLSolversBase then it should be reflected here "automatically" (we may need to introduce some keywords for ease of use)
